rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,325 @@
+
+### Config Changes
+
+Encrypted session recording is a feature of the auth service and can be enabled
+through the `session_recording_config` resource.
+```yaml
+# session_recording_config.yml
+kind: session_recording_config
+version: v2
+spec:
+  encrypted: true
+```
+HSM integration is facilitated through the existing configuration
+options for setting up an HSM backed CA keystore through pkcs#11. Example
+configuration found [here](https://goteleport.com/docs/admin-guides/deploy-a-cluster/hsm/#step-25-configure-teleport).
+
+### Protobuf Changes
+```proto
+// api/proto/teleport/legacy/types/types.proto
+
+// WrappedKey wraps a PrivateKey using an asymmetric keypair.
+message WrappedKey {
+  // WrappingPair is the asymmetric keypair used to wrap the private key.
+  // Expected to be RSA
+  SSHKeyPair WrappingPair = 1;
+  bytes WrappedPrivateKey = 2 [
+    (gogoproto.nullable) = true, // must be nullable
+    (gogoproto.jsontag) = "wrapped_private_key"
+  ];
+  // The public key is included unencrypted to make it easier to find rotated
+  // keypairs and to make available to proxy and host nodes.
+  bytes PublicKey = 3 [(gogoproto.jsontag) = "public_key"];
+  // Signals that a key should be rotated
+  bool rotate = 4 [(gogoproto.jsontag) = "rotate"];
+}
+
+// SessionRecordingConfigStatusV2 contains all of the current and rotated keys
+// used for encrypted session recording.
+message SessionRecordingConfigStatusV2 {
+  // ActiveKeys is a list of active, wrapped X25519 private keys. There should
+  // be at most one wrapped key per auth server using the
+  // SessionRecordingConfigV2 resource unless keys are being rotated.
+  repeated WrappedKey ActiveKeys = 2 [
+    (gogoproto.jsontag) = "active_keys"
+  ];
+  // RotatedKeys is a list of wrapped private keys that have been rotated.
+  // These are kept to decrypt historical encrypted session recordings.
+  repeated WrappedKey RotatedKeys = 3 [
+    (gogoproto.jsontag) = "rotated_keys"
+  ];
+}
+
+// SessionRecordingConfigSpecV2 is the actual data we care about
+// for SessionRecordingConfig.
+message SessionRecordingConfigSpecV2 {
+  // existing fields omitted 
+
+  BoolValue encrypted = 3 [(gogoproto.jsontag) = "encrypted"]
+}
+
+// SessionRecordingConfigV2 contains session recording configuration.
+message SessionRecordingConfigV2 {
+  // existing fields omitted
+
+  // Status contains all of the current and rotated keys used for encrypted
+  // session recording
+  SessionRecordingConfigStatusV2 Status = 6 [
+    (gogoproto.nullable) = true,
+    (gogoproto.jsontag) = "status"
+  ]
+}
+
+// api/proto/teleport/trust/v1/trust_service.proto
+
+// TrustService provides methods to manage certificate authorities.
+service TrustService {
+  // existing RPCs omitted
+
+  // RotateSessionRecordingKeys rotates the keys associated with encrypting and
+  // decrypting session recordings.
+  rpc RotateSessionRecordingKeys(RotateSessionRecordingKeysRequest) returns (RotateSessionRecordingKeysResponse);
```

Comment:
I went back and forth on whether or not this was the right place for this. I'm not sure it warrants a new service, but I'm certainly open to feedback here.